API Introduce hooks to help components figure out what tab they are in. (#1129)
silverstripe/silverstripe-admin: client/src/components/Tabs/README.md

 * `tabClassName` (string): Class to use for the tab.

 _NOTE:_ For other properties, please refer to the [reactstrap Tabs](https://reactstrap.github.io/components/tabs/) documentation.

# Interacting with Tabs from a child component

> Components inside a tab can retrieve information about their parent using these method:

Author: `injectTabContext` is an HOC. It could be used with a functional component as well. Not sure there's much point dividing up a list of three items.

> Components inside a tab can retrieve information about their parent using these method:

Reviewer: I think there is, basically

If you have a functional component then do this:
...

If you have a class component then do this:
...

Doesn't seem like much extra work to make the instructions better?

> Components inside a tab can retrieve information about their parent using these method:

Author: It's not clear to me this makes the doc better. You could argue that `injectTabContext` and `useTabContext` should be together  because they essentially do the same thing and `useTabFirstShow` should be on its own.

At the of of the day, these 3 methods do related things. I think it makes sense to discuss them as a whole.

> - `useTabContext` which is a hook returning the tab context for functional components

Author: `useTabContext` is a custom react hook. The react convention is to call hooks `use` something (e.g. `useState`, `useEffect`, `useContext`).

- `injectTabContext` which injects the tab context as props in your class component

> - `useTabFirstShow` which is a hook that allows you to execute a callback when a tab is first shown.

Author: Same deal ... it's a hook.

The `tabContext` returned by `useTabContext` and `injectTabContext` contains the following properties
- `activeTab` (string): name of the active tab in the nearest tab set
- `currentTab` (string): name of the tab the current component resides in
- `isOnActiveTab` (boolean): whether the current component is in a tab that is currently visible.

The `tabContext` will be `false` for components outside of a tab.
 

## Example of `useTabContext`

```jsx
import React from 'react';
import useTabContext from 'hooks/useTabContext';

export function MyComponent() {
  // useTabContext only works inside functinal components
  const tabContext = useTabContext();
  if (!tabContext) {
    return <p>I'm not in a tab.</p>;
  }

  const {activeTab, currentTab, isOnActiveTab} = tabContext;
  

>   return 

Author: Brackets are not necessary here. This should work as is.

>   return 

Reviewer: Tried locally, PHPstorm complains
![image](https://user-images.githubusercontent.com/4809037/95147608-b45b9600-07dd-11eb-9d98-51c003e47325.png)

No complaint with brackets

![image](https://user-images.githubusercontent.com/4809037/95147629-c2a9b200-07dd-11eb-9d3e-0c1053852d84.png)

    <p>
        The {activeTab} tab is currently active.

>         I'm on the {currentTab} tab which is {isOnActiveTab ? 'visible' : 'invisible'}.
>     </p>; 

Reviewer: ```suggestion
        I&apos;m on the {currentTab} tab which is {isOnActiveTab ? 'visible' : 'invisible'}.
    </p>; 
```

}
```

## Example of `injectTabContext`

```jsx
import React, {Component} from 'react';
import {injectTabContext} from 'hooks/useTabContext';

class MyComponent extends Component {

>   render({tabContext}) {  

Author: `tabContext` would be a prop here. So it would come in as a property that need destructuring.

    if (!tabContext) {
      return <p>I'm not in a tab.</p>;
    }
    
    const {activeTab, currentTab, isOnActiveTab} = tabContext;
      
    return 
      <p>
        The {activeTab} tab is currently active.
        I'm on the {currentTab} tab which is {isOnActiveTab ? 'visible' : 'invisible'}.
      </p>;
  }
}

// `injectTabContext` is a higher-order-component (HOC). 

> // It can be applied to class and functional components.

Reviewer: Seems like we'd only want to use it on class components though.  Kind of confusing saying we could use it on functional components, meanwhile we have 2 other functional specifically for functional components

Author: Some people prefer to use component composition. To be honest, the only reason I created `injectTabContext` was because `provideUsedOnData` is a class component. I would have preferred to just use hooks.

Reviewer: How about

`// It is used to add context to class components.  It can also be used on functional components.`

export default injectTabContext(MyComponent);
```

## Example of `useTabFirstShow`

```jsx
import React from 'react';
import {useTabFirstShow} from 'hooks/useTabContext';

export function MyComponent({loadDataAction}) {
  // `useTabFirstShow` is a Silverstripe specific hook. It will triger the provided function once 
  // the component is visible. The tabContext is passed back to the provided method.
  // `useTabFirstShow` can be usefull to delay an expensive request until it's actually needed.
  // Because `useTabFirstShow` is a hook, it can only be called from a functional component.

>   // Use `injectTabContext` and lifecyle methods to achieve the same result in a class component.

Author: The doc assumes some degree of familiarity react. Don't really want to be giving a react crash course here. We could link off to the main react doc.

The `tabContext` returned by `useTabContext` and `injectTabContext` contains the following properties
- `activeTab` (string): name of the active tab in the nearest tab set
- `currentTab` (string): name of the tab the current component resides in
- `isOnActiveTab` (boolean): whether the current component is in a tab that is currently visible.

The `tabContext` will be `false` for components outside of a tab.
 

## Example of `useTabContext`

```jsx
import React from 'react';
import useTabContext from 'hooks/useTabContext';

export function MyComponent() {
  // useTabContext only works inside functinal components
  const tabContext = useTabContext();
  if (!tabContext) {
    return <p>I'm not in a tab.</p>;
  }

  const {activeTab, currentTab, isOnActiveTab} = tabContext;
  
  return (
    <p>
        The {activeTab} tab is currently active.
        I'm on the {currentTab} tab which is {isOnActiveTab ? 'visible' : 'invisible'}.
    </p>
  ); 
}
```

## Example of `injectTabContext`

```jsx
import React, {Component} from 'react';
import {injectTabContext} from 'hooks/useTabContext';

class MyComponent extends Component {

>   render({tabContext}) {  

Reviewer: I'm not sure the `render({tabContext}) {` is correct

It threw errors on my local

However this worked as expected

```
render() {
  this.props.tabContext;
```

Author: You're right ... I've been doing too much functional components.

 * `tabClassName` (string): Class to use for the tab.

 _NOTE:_ For other properties, please refer to the [reactstrap Tabs](https://reactstrap.github.io/components/tabs/) documentation.

# Interacting with Tabs from a child component

Components inside a tab can retrieve information about their parent using these method:
- `useTabContext` which is a hook returning the tab context for functional components
- `injectTabContext` which injects the tab context as props in your class component
- `useTabFirstShow` which is a hook that allows you to execute a callback when a tab is first shown.

The `tabContext` returned by `useTabContext` and `injectTabContext` contains the following properties
- `activeTab` (string): name of the active tab in the nearest tab set
- `currentTab` (string): name of the tab the current component resides in
- `isOnActiveTab` (boolean): whether the current component is in a tab that is currently visible.

The `tabContext` will be `false` for components outside of a tab.
 

## Example of `useTabContext`

```jsx
import React from 'react';
import useTabContext from 'hooks/useTabContext';

export function MyComponent() {

>   // useTabContext only works inside functinal components

Reviewer: ```suggestion
  // useTabContext only works inside functional components
```

  const tabContext = useTabContext();
  if (!tabContext) {
    return <p>I'm not in a tab.</p>;
  }

  const { activeTab, currentTab, isOnActiveTab } = tabContext;
  
  return (
    <p>
        The {activeTab} tab is currently active.
        I'm on the {currentTab} tab which is {isOnActiveTab ? 'visible' : 'invisible'}.
    </p>
  ); 
}
```

## Example of `injectTabContext`

```jsx
import React, { Component } from 'react';
import { injectTabContext } from 'hooks/useTabContext';

class MyComponent extends Component {
  render() {  

>     const {tabContext} = this.props;

Reviewer: ```suggestion
    const { tabContext } = this.props;
```

    if (!tabContext) {
      return <p>I'm not in a tab.</p>;
    }
    
    const { activeTab, currentTab, isOnActiveTab } = tabContext;
      
    return (
      <p>
        The {activeTab} tab is currently active.
        I'm on the {currentTab} tab which is {isOnActiveTab ? 'visible' : 'invisible'}.
      </p>
    );
  }
}

// `injectTabContext` is a higher-order-component (HOC). It is used to add context to class 
// components. It can also be used on functional components.
export default injectTabContext(MyComponent);
```

## Example of `useTabFirstShow`

```jsx
import React from 'react';
import { useTabFirstShow } from 'hooks/useTabContext';

> export function MyComponent({loadDataCallback}) {

Reviewer: ```suggestion
export function MyComponent({ loadDataCallback }) {
```
